File: backend/app/core/permissions.py
from __future__ import annotations

from enum import StrEnum


class Permission(StrEnum):
    COMPANY_MANAGE = "company_manage"
    USER_MANAGE = "user_manage"
    TEMPLATE_MANAGE = "template_manage"
    CONVERSION_PROCESS = "conversion_process"
    CONVERSION_CONFIRM = "conversion_confirm"
    EXPORT_RUN = "export_run"
    RULE_APPROVE = "rule_approve"
    AUDIT_VIEW = "audit_view"
    READ = "read"


ROLE_PERMISSIONS: dict[str, set[Permission]] = {
    "admin": set(Permission),
    "template_admin": {Permission.READ, Permission.TEMPLATE_MANAGE},
    "processor": {
        Permission.READ,
        Permission.CONVERSION_PROCESS,
        Permission.CONVERSION_CONFIRM,
        Permission.EXPORT_RUN,
    },
    "reviewer": {
        Permission.READ,
        Permission.CONVERSION_CONFIRM,
        Permission.RULE_APPROVE,
        Permission.AUDIT_VIEW,
    },
    "auditor": {Permission.READ, Permission.AUDIT_VIEW},
}

CROSS_COMPANY_ROLES: frozenset[str] = frozenset({"admin", "auditor"})


def permissions_for(roles: list[str]) -> set[Permission]:
    perms: set[Permission] = set()
    for role in roles:
        perms |= ROLE_PERMISSIONS.get(role, set())
    return perms

File: frontend/src/auth/permissions.ts
/**
 * 镜像后端角色→权限映射（backend/app/core/enums.py Role → Permission）。
 * 权限值与后端 Permission enum value 完全一致，修改时需同步后端。
 */
const ROLE_PERMS: Record<string, string[]> = {
  admin: [
    'company_manage',
    'user_manage',
    'template_manage',
    'conversion_process',
    'conversion_confirm',
    'export_run',
    'rule_approve',
    'audit_view',
    'read',
  ],
  template_admin: ['read', 'template_manage'],
  processor: ['read', 'conversion_process', 'conversion_confirm', 'export_run'],
  reviewer: ['read', 'conversion_confirm', 'rule_approve', 'audit_view'],
  auditor: ['read', 'audit_view'],
};

/** 合并多个角色对应的权限集合。 */
export function permissionsForRoles(roles: string[]): Set<string> {
  const s = new Set<string>();
  roles.forEach((r) => (ROLE_PERMS[r] ?? []).forEach((p) => s.add(p)));
  return s;
}
